src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ModelBinderFactory.cs

```diff
@@ -184,15 +221,33 @@ private class DefaultModelBinderProviderContext : ModelBinderProviderContext
             public override IModelMetadataProvider MetadataProvider { get; }
 
             // Not using a 'real' Stack<> because we want random access to modify the entries.
-            public List<KeyValuePair<Key, PlaceholderBinder>> Stack { get; }
+            public Dictionary<Key, IModelBinder> Collection { get; }
 
             public override IModelBinder CreateBinder(ModelMetadata metadata)
             {
+                if (metadata == null)
+                {
+                    throw new ArgumentNullException(nameof(metadata));
+                }
+
+                // For interior nodes we use the ModelMetadata as the cache token. This ensures that all interior
+                // nodes with the same metadata will have the the same binder. This is OK because for an interior
+                // node there's no opportunity to customize binding info like there is for a parameter.
+                var token = metadata;
+
                 var nestedContext = new DefaultModelBinderProviderContext(this, metadata);
-                return _factory.CreateBinderCore(nestedContext, token: null);
+                return _factory.CreateBinderCoreCached(nestedContext, token);
             }
         }
 
+        // This key allows you to specify a ModelMetadata which represents the type/property being bound
+        // and a 'token' which acts as an arbitrary discriminator.
+        //
+        // This is necessary because the same metadata might be bound as a top-level parameter (with BindingInfo on
+        // the ParameterDescriptor) or in a call to TryUpdateModel (no BindingInfo) or as a collection element.
+        //
+        // We need to be able to tell the difference between these things to avoid over-caching.
+        [DebuggerDisplay("{ToString(),nq}")]
```

Comment:
The debugger should fall back to the `ToString()` method w/o this attribute. Either remove this or rename the method and make it `private`.